src/contracts/core/DelegationManager.sol

```diff
@@ -224,48 +227,171 @@ contract DelegationManager is Initializable, OwnableUpgradeable, Pausable, Deleg
+                "DelegationManager.completeQueuedAction: input length mismatch"
+            );
+        }
+
+        // Remove `withdrawalRoot` from pending roots
+        delete pendingWithdrawals[withdrawalRoot];
+
+        address currentOperator = delegatedTo[msg.sender];
+
+        // Finalize action by converting shares to tokens for each strategy, or
+        // by re-awarding shares in each strategy.
+        for (uint256 i = 0; i < withdrawal.strategies.length; ) {
+            if (receiveAsTokens) {
+                _withdrawSharesAsTokens({
+                    staker: withdrawal.staker,
+                    withdrawer: msg.sender,
+                    strategy: withdrawal.strategies[i],
+                    shares: withdrawal.shares[i],
+                    token: tokens[i]
+                });
+            } else {
+                // Award shares back to staker in StrategyManager/EigenPodManager
+                // If staker is delegated, increases shares delegated to operator
+                _addAndDelegateShares({
+                    staker: withdrawal.staker,
+                    withdrawer: msg.sender,
+                    operator: currentOperator,
+                    strategy: withdrawal.strategies[i],
+                    shares: withdrawal.shares[i]
+                });
+            }
+
+            unchecked { ++i; }
+        }
+
+        emit WithdrawalCompleted(withdrawalRoot);
+    }
+
+    /// @notice Migrates an existing queued withdrawal from the StrategyManager contract to this contract.
+    /// @dev This function is expected to be removed in the next upgrade, after all queued withdrawals have been migrated.
+    function migrateQueuedWithdrawal(IStrategyManager.DeprecatedStruct_QueuedWithdrawal memory strategyManagerWithdrawalToMigrate) external {
+        // check for existence and delete the old storage
+        bytes32 oldWithdrawalRoot = strategyManager.calculateWithdrawalRoot(strategyManagerWithdrawalToMigrate);
+        strategyManager.migrateQueuedWithdrawal(oldWithdrawalRoot);
+
+        address staker = strategyManagerWithdrawalToMigrate.depositor;
+        // Create queue entry and increment withdrawal nonce
+        uint256 nonce = cumulativeWithdrawalsQueued[staker];
+        cumulativeWithdrawalsQueued[staker]++;
+
+        Withdrawal memory migratedWithdrawal = Withdrawal({
+            staker: staker,
+            delegatedTo: strategyManagerWithdrawalToMigrate.delegatedAddress,
+            withdrawer: strategyManagerWithdrawalToMigrate.withdrawerAndNonce.withdrawer,
+            nonce: nonce,
+            startBlock: strategyManagerWithdrawalToMigrate.withdrawalStartBlock,
+            strategies: strategyManagerWithdrawalToMigrate.strategies,
+            shares: strategyManagerWithdrawalToMigrate.shares
+        });
+
+        // create the new storage
+        bytes32 newRoot = calculateWithdrawalRoot(migratedWithdrawal);
+        pendingWithdrawals[newRoot] = true;
```

Comment:
Just for sanity's sake, we should require that the new calculated root is "not" currently pending. Just in case.

Reply:
sure, it's basically an assertion, since it really, really should never happen.
added here https://github.com/Layr-Labs/eigenlayer-contracts/pull/222/commits/726aceb6e9eb320ece7057b432948e0c91f6f471